pkg/controller/drain.go

```diff
@@ -373,6 +401,313 @@ func (o *DrainOptions) evictPods(pods []api.Pod, policyGroupVersion string, getP
 	return utilerrors.NewAggregate(errors)
 }
 
+func (o *DrainOptions) evictPodsWithoutPv(pods []*corev1.Pod, policyGroupVersion string, getPodFn func(namespace, name string) (*api.Pod, error), returnCh chan error) {
+	for _, pod := range pods {
+		go o.evictPodInternal(pod, policyGroupVersion, getPodFn, returnCh)
+	}
+	return
+}
+
+func sortPodsByPriority(pods []*corev1.Pod) {
+	sort.Slice(pods, func(i, j int) bool {
+		return *pods[i].Spec.Priority > *pods[j].Spec.Priority
+	})
+}
+
+// doAccountingOfPvs returns the map with keys as pod names and values as array of attached volumes' IDs
+func (o *DrainOptions) doAccountingOfPvs(pods []*corev1.Pod) map[string][]string {
+	volMap := make(map[string][]string)
+	pvMap := make(map[string][]string)
+
+	for _, pod := range pods {
+		podPVs, _ := o.getPvs(pod)
+		pvMap[pod.Namespace+"/"+pod.Name] = podPVs
+	}
+	glog.V(4).Info("PV map: ", pvMap)
+
+	filterSharedPVs(pvMap)
+
+	for i := range pvMap {
+		pvList := pvMap[i]
+		vols, err := o.getVolIDsFromDriver(pvList)
+		if err != nil {
+			// In case of error, log and skip this set of volumes
+			glog.Errorf("Error getting volume ID from cloud provider. Skipping volumes for pod: %v. Err: %v", i, err)
+			continue
+		}
+		volMap[i] = vols
+	}
+	glog.V(4).Info("Volume map: ", volMap)
+	return volMap
+}
+
+// filterSharedPVs filters out the PVs that are shared among pods.
+func filterSharedPVs(pvMap map[string][]string) {
+	sharedVol := make(map[string]bool)
+	sharedVolumesFound := false
+
+	// Create hash map of volumes:
+	// Key: volume name
+	// Value: 'true' if any other pod shares this volume, else 'false'
+	for _, vols := range pvMap {
+		for _, vol := range vols {
+			if _, ok := sharedVol[vol]; !ok {
+				sharedVol[vol] = false
+			} else {
+				sharedVol[vol] = true
+				sharedVolumesFound = true
+			}
+		}
+	}
+
+	if !sharedVolumesFound {
+		glog.V(4).Info("No shared volumes found.")
+		return
+	}
+
+	// Recreate the values of pvMap. Append volume if it is not shared
+	for pod, vols := range pvMap {
+		volList := []string{}
+		for _, vol := range vols {
+			if sharedVol[vol] == false {
+				volList = append(volList, vol)
+			}
+		}
+		pvMap[pod] = volList
+	}
+	glog.V(4).Info("Removed shared volumes. Filtered list of pods with volumes: ", pvMap)
+}
+
+func (o *DrainOptions) evictPodsWithPv(pods []*corev1.Pod, policyGroupVersion string, getPodFn func(namespace, name string) (*api.Pod, error), returnCh chan error) {
+	detachCh := make(chan error, 1)
```

Comment:
Closing the channel before making sure that all the go routines writing to this channel are done with their task will lead to panic.
As discussed, `waitForDetach()` function is now called in-line instead of as a go routine, and moved the timeout check inside the function.